pkg/kv/kvserver/replica_evaluate.go

```diff
@@ -333,7 +337,16 @@ func evaluateBatch(
 		if limit := baHeader.MaxSpanRequestKeys; limit > 0 {
 			retResults := reply.Header().NumKeys
 			if retResults > limit {
-				log.Fatalf(ctx, "received %d results, limit was %d", retResults, limit)
+				index, retResults, limit := index, retResults, limit // don't alloc unless branch taken
+				err := errorutil.UnexpectedWithIssueErrorf(46652,
+					"received %d results, limit was %d (original limit: %d, batch=%s idx=%d)",
+					errors.Safe(retResults), errors.Safe(limit),
+					errors.Safe(ba.Header.MaxSpanRequestKeys), // already on heap
+					errors.Safe(ba.Summary()), errors.Safe(index))
+				if sentryIssue46720Limiter.Allow() {
```

Comment:
@knz just curious if you think this is how this pattern of "user error + report" should be done.